File: research/experiments/sweep/runs/opt-large-go-sonnet45/finite_diff_test.go
package optimization

import (
	"math"
	"testing"
)

func TestForwardDiffGradientSphere(t *testing.T) {
	x := []float64{3, 4}
	grad := ForwardDiffGradient(Sphere, x)
	expected := SphereGrad(x) // [6, 8]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-7 {
			t.Errorf("ForwardDiffGradient(Sphere, [3,4])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}

	// Test at origin
	x = []float64{0, 0}
	grad = ForwardDiffGradient(Sphere, x)
	expected = SphereGrad(x) // [0, 0]
	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-7 {
			t.Errorf("ForwardDiffGradient(Sphere, [0,0])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestCentralDiffGradientSphere(t *testing.T) {
	x := []float64{3, 4}
	grad := CentralDiffGradient(Sphere, x)
	expected := SphereGrad(x) // [6, 8]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-10 {
			t.Errorf("CentralDiffGradient(Sphere, [3,4])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}

	// Test at origin
	x = []float64{0, 0}
	grad = CentralDiffGradient(Sphere, x)
	expected = SphereGrad(x) // [0, 0]
	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-10 {
			t.Errorf("CentralDiffGradient(Sphere, [0,0])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestForwardDiffGradientRosenbrock(t *testing.T) {
	x := []float64{-1.2, 1.0}
	grad := ForwardDiffGradient(Rosenbrock, x)
	expected := RosenbrockGrad(x) // [-215.6, -88]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-4 {
			t.Errorf("ForwardDiffGradient(Rosenbrock, [-1.2,1.0])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestCentralDiffGradientRosenbrock(t *testing.T) {
	x := []float64{-1.2, 1.0}
	grad := CentralDiffGradient(Rosenbrock, x)
	expected := RosenbrockGrad(x) // [-215.6, -88]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-7 {
			t.Errorf("CentralDiffGradient(Rosenbrock, [-1.2,1.0])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestForwardDiffGradientBeale(t *testing.T) {
	x := []float64{1, 1}
	grad := ForwardDiffGradient(Beale, x)
	expected := BealeGrad(x) // [-1.5, 5.25]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-5 {
			t.Errorf("ForwardDiffGradient(Beale, [1,1])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestCentralDiffGradientBeale(t *testing.T) {
	x := []float64{1, 1}
	grad := CentralDiffGradient(Beale, x)
	expected := BealeGrad(x) // [-1.5, 5.25]

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-8 {
			t.Errorf("CentralDiffGradient(Beale, [1,1])[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

func TestMakeGradient(t *testing.T) {
	// Test forward (default)
	gradFunc := MakeGradient(Sphere, "forward")
	x := []float64{3, 4}
	grad := gradFunc(x)
	expected := SphereGrad(x)

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-7 {
			t.Errorf("MakeGradient(forward)[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}

	// Test central
	gradFunc = MakeGradient(Sphere, "central")
	grad = gradFunc(x)

	for i := range grad {
		if math.Abs(grad[i]-expected[i]) > 1e-10 {
			t.Errorf("MakeGradient(central)[%d] = %v, want %v", i, grad[i], expected[i])
		}
	}
}

File: research/experiments/sweep/runs/opt-medium-kotlin-sonnet45/src/test/kotlin/optimization/FiniteDiffTest.kt
package optimization

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.math.abs

class FiniteDiffTest {
    // Test functions with analytic gradients
    private fun sphere(x: DoubleArray): Double = x.sumOf { it * it }
    private fun sphereGrad(x: DoubleArray): DoubleArray = DoubleArray(x.size) { 2.0 * x[it] }

    private fun rosenbrock(x: DoubleArray): Double {
        val a = 1.0 - x[0]
        val b = x[1] - x[0] * x[0]
        return a * a + 100.0 * b * b
    }

    private fun rosenbrockGrad(x: DoubleArray): DoubleArray {
        val g = DoubleArray(2)
        g[0] = -2.0 * (1.0 - x[0]) - 400.0 * x[0] * (x[1] - x[0] * x[0])
        g[1] = 200.0 * (x[1] - x[0] * x[0])
        return g
    }

    private fun beale(x: DoubleArray): Double {
        val t1 = 1.5 - x[0] + x[0] * x[1]
        val t2 = 2.25 - x[0] + x[0] * x[1] * x[1]
        val t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1]
        return t1 * t1 + t2 * t2 + t3 * t3
    }

    private fun bealeGrad(x: DoubleArray): DoubleArray {
        val t1 = 1.5 - x[0] + x[0] * x[1]
        val t2 = 2.25 - x[0] + x[0] * x[1] * x[1]
        val t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1]
        val g = DoubleArray(2)
        g[0] = 2.0 * t1 * (-1.0 + x[1]) + 2.0 * t2 * (-1.0 + x[1] * x[1]) + 2.0 * t3 * (-1.0 + x[1] * x[1] * x[1])
        g[1] = 2.0 * t1 * x[0] + 2.0 * t2 * (2.0 * x[0] * x[1]) + 2.0 * t3 * (3.0 * x[0] * x[1] * x[1])
        return g
    }

    private fun assertArrayEquals(expected: DoubleArray, actual: DoubleArray, tol: Double) {
        assertEquals(expected.size, actual.size)
        for (i in expected.indices) {
            assertEquals(expected[i], actual[i], tol, "Mismatch at index $i")
        }
    }

    @Test
    fun testForwardDiffSphere3_4() {
        val x = doubleArrayOf(3.0, 4.0)
        val expected = sphereGrad(x)
        val actual = forwardDiffGradient(::sphere, x)
        assertArrayEquals(expected, actual, 1e-6)
    }

    @Test
    fun testForwardDiffSphere0_0() {
        val x = doubleArrayOf(0.0, 0.0)
        val expected = sphereGrad(x)
        val actual = forwardDiffGradient(::sphere, x)
        assertArrayEquals(expected, actual, 1e-7)
    }

    @Test
    fun testForwardDiffRosenbrock() {
        val x = doubleArrayOf(-1.2, 1.0)
        val expected = rosenbrockGrad(x)
        val actual = forwardDiffGradient(::rosenbrock, x)
        assertArrayEquals(expected, actual, 1e-4)
    }

    @Test
    fun testForwardDiffBeale() {
        val x = doubleArrayOf(1.0, 1.0)
        val expected = bealeGrad(x)
        val actual = forwardDiffGradient(::beale, x)
        assertArrayEquals(expected, actual, 1e-5)
    }

    @Test
    fun testCentralDiffSphere3_4() {
        val x = doubleArrayOf(3.0, 4.0)
        val expected = sphereGrad(x)
        val actual = centralDiffGradient(::sphere, x)
        assertArrayEquals(expected, actual, 1e-10)
    }

    @Test
    fun testCentralDiffSphere0_0() {
        val x = doubleArrayOf(0.0, 0.0)
        val expected = sphereGrad(x)
        val actual = centralDiffGradient(::sphere, x)
        assertArrayEquals(expected, actual, 1e-10)
    }

    @Test
    fun testCentralDiffRosenbrock() {
        val x = doubleArrayOf(-1.2, 1.0)
        val expected = rosenbrockGrad(x)
        val actual = centralDiffGradient(::rosenbrock, x)
        assertArrayEquals(expected, actual, 1e-7)
    }

    @Test
    fun testCentralDiffBeale() {
        val x = doubleArrayOf(1.0, 1.0)
        val expected = bealeGrad(x)
        val actual = centralDiffGradient(::beale, x)
        assertArrayEquals(expected, actual, 1e-8)
    }

    @Test
    fun testMakeGradientForward() {
        val gradFn = makeGradient(::sphere)
        val x = doubleArrayOf(3.0, 4.0)
        val expected = forwardDiffGradient(::sphere, x)
        val actual = gradFn(x)
        assertArrayEquals(expected, actual, 1e-15)
    }

    @Test
    fun testMakeGradientCentral() {
        val gradFn = makeGradient(::sphere, "central")
        val x = doubleArrayOf(3.0, 4.0)
        val expected = centralDiffGradient(::sphere, x)
        val actual = gradFn(x)
        assertArrayEquals(expected, actual, 1e-15)
    }
}
